## Практическая работа

В этой работе вам нужно будет подготовить финальный нотбук по задаче предсказания стоимости поддержанного автомобиля, которую мы решали на протяжении нескольких модулей. 
<br>
*Обязательные задачи* помогут проверить, что вы действительно усвоили материал. Если у вас всё получилось, можете переходить к следующей теме.
<br>
Удачи!

Цели практической работы: 

1.   Потренироваться в формировании понятного и структурированного ноутбука по результатам исследования.
2.   Отработать выбор наилучшей предсказательной модели с помощью кросс-валидации. 
3.   На основе результатов метрик научиться делать выводы о переобучении/недообучении модели.
4.   Научиться сохранять обученные модели в формате `pickle`.




Что оценивается:

*   Ноутбук состоит из блоков: каждый блок соответствует определённой фазе работы с задачей.
*   Блоки, которые состоят из нескольких этапов, декомпозированы на подблоки (например, в `Data Preparation` может быть несколько подблоков: `Data cleaning`, `Feature Selection` и тому подобное).
*   Код написан в едином стиле и не нарушает PEP-8.
*   Код покрыт комментариями.
*   Код воспроизводим (то есть проверяющий, запустив ноутбук, получит ровно такие же результаты).
*   Все условия задач выполнены.





## Обязательные задачи

### Постановка общей задачи:



Напомним, что вы работаете с небольшой выборкой из коллекции подержанных автомобилей, выставленных на продажу в Соединённых Штатах. На этих данных вам нужно построить модель классификации, определяющую категорию цены подержанного автомобиля в зависимости от характеристик транспортного средства.

### Описание датасета:



- `id`: идентификатор записи;
- `url`: URL записи о продаже;
- `region`: регион;
- `region_url`: URL региона;
- `price`: стоимость;
- `year`: год выпуска;
- `manufacturer`: производитель;
- `model`: модель;
- `condition`: состояние;
- `cylinders`: количество цилиндров;
- `fuel`: тип топлива;
- `odometer`: количество пройденных миль;
- `title_status`: статус;
- `transmission`: коробка передач;
- `VIN`: идентификационный номер;
- `drive`: тип привода;
- `size`: размер;
- `type`: кузов;
- `paint_color`: цвет;
- `image_url`: URL изображения;
- `description`: указанное описание;
- `county`: страна;
- `state`: штат;
- `lat`: широта;
- `long`: долгота;
- `posting_date`: дата размещения объявления о продаже;
- `price_category`: категория цены.

###  Задачи:

Ниже представлены задачи по проекту построения модели. Выполните их и в отдельном файле соберите финальный ноутбук проекта. 

## 0. *Импорт сторонних библиотек*

В первой ячейке ноутбука импортируйте библиотеки, которые необходимы для работы с данными. Впоследствии все импорты добавляйте в эту ячейку. Структурируйте их по PEP-8. 

[Документация по оформлению import'ов (PEP-8)](https://peps.python.org/pep-0008/#imports)

In [163]:
import pandas as pd

from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, cross_validate, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_validate, cross_val_score

## 1. *Загрузка данных*

Загрузите данные из файла `df_out`. Выведите размерность и первые 5 строчек данных.

In [164]:
# Загрузим датасет выведем первые 5 строк
df = pd.read_csv('data/df_out.csv')
df.head()

id                                                url  \
0  7308295377  https://chattanooga.craigslist.org/ctd/d/chatt...   
1  7316380095  https://newjersey.craigslist.org/ctd/d/carlsta...   
2  7313733749  https://reno.craigslist.org/ctd/d/atlanta-2017...   
3  7308210929  https://fayetteville.craigslist.org/ctd/d/rale...   
4  7316474668  https://newyork.craigslist.org/lgi/cto/d/baldw...   

          region                           region_url  price    year  \
0    chattanooga   https://chattanooga.craigslist.org  54990  2020.0   
1   north jersey     https://newjersey.craigslist.org  16942  2016.0   
2   reno / tahoe          https://reno.craigslist.org  35590  2017.0   
3   fayetteville  https://fayetteville.craigslist.org  14500  2013.0   
4  new york city       https://newyork.craigslist.org  21800  2021.0   

  manufacturer                   model    fuel  odometer title_status  \
0          ram  2500 crew cab big horn  diesel     27442        clean   
1         ford    explorer 4wd 4dr xlt     NaN     60023        clean   
2   volkswagen        golf r hatchback     gas     14048        clean   
3       toyota                    rav4     gas    117291        clean   
4       nissan                  altima     gas      8000        clean   

  transmission                                          image_url  \
0        other  https://images.craigslist.org/00N0N_1xMPvfxRAI...   
1    automatic  https://images.craigslist.org/00x0x_26jl9F0cnL...   
2        other  https://images.craigslist.org/00y0y_eeZjWeiSfb...   
3    automatic  https://images.craigslist.org/00606_iGe5iXidib...   
4    automatic  https://images.craigslist.org/00V0V_3pSOiPZ3Sd...   

                                         description state        lat  \
0  Carvana is the safer way to buy a car During t...    tn  35.060000   
1  ***Call Us for more information at: 201-635-14...    nj  40.821805   
2  Carvana is the safer way to buy a car During t...    ca  33.779214   
3  2013 Toyota RAV4 XLE 4dr SUV     Offered by: R...    nc  35.715954   
4  2021 Nissan Altima Sv with Only 8 K Miles Titl...    ny  40.654800   

        long              posting_date price_category  \
0 -85.250000  2021-04-17T12:30:50-0400           high   
1 -74.061962  2021-05-03T15:40:21-0400         medium   
2 -84.411811  2021-04-28T03:52:20-0700           high   
3 -78.655304  2021-04-17T10:08:57-0400         medium   
4 -73.609700  2021-05-03T18:32:06-0400         medium   

                        date  
0  2021-04-17 16:30:50+00:00  
1  2021-05-03 19:40:21+00:00  
2  2021-04-28 10:52:20+00:00  
3  2021-04-17 14:08:57+00:00  
4  2021-05-03 22:32:06+00:00

## 2. *Data Preparation*

Объявите блок Data Preparation. В этом блоке:
* произведите преобразование типов данных, если нужно;
* исследуйте данные на пропуски, обработайте их (например, заполните какими-то значениями);
* избавьтесь от аномалий, если нужно.

В решении данной задачи опирайтесь на результаты работы в 17-19 модулях блока Data Preparation. Исследование данных подкрепляйте соответствующей визуализацией в виде графиков, таблиц или статистических значений.

Декомпозируйте блок на подблоки.


### Data type conversion

In [165]:
df.shape

(10000, 20)

In [166]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              10000 non-null  int64  
 1   url             10000 non-null  object 
 2   region          10000 non-null  object 
 3   region_url      10000 non-null  object 
 4   price           10000 non-null  int64  
 5   year            9964 non-null   float64
 6   manufacturer    10000 non-null  object 
 7   model           9872 non-null   object 
 8   fuel            9937 non-null   object 
 9   odometer        10000 non-null  int64  
 10  title_status    9834 non-null   object 
 11  transmission    9955 non-null   object 
 12  image_url       9998 non-null   object 
 13  description     9998 non-null   object 
 14  state           10000 non-null  object 
 15  lat             9902 non-null   float64
 16  long            9902 non-null   float64
 17  posting_date    9998 non-null   

In [167]:
df.dtypes

id                  int64
url                object
region             object
region_url         object
price               int64
year              float64
manufacturer       object
model              object
fuel               object
odometer            int64
title_status       object
transmission       object
image_url          object
description        object
state              object
lat               float64
long              float64
posting_date       object
price_category     object
date               object
dtype: object

---

date            9998 non-null, две строки с пропусками, найдем и проанализируем эти строки

---

In [168]:
df[df.date.isnull()]

id                                                url  \
6728  7315259025  https://nh.craigslist.org/cto/d/manchester-201...   
9388  7208549803  https://bellingham.craigslist.org/ctd/d/bellin...   

             region                         region_url  price  year  \
6728  new hampshire          https://nh.craigslist.org   3750   NaN   
9388     bellingham  https://bellingham.craigslist.org  11999   NaN   

     manufacturer model fuel  odometer title_status transmission image_url  \
6728        other   NaN  NaN     95674          NaN          NaN       NaN   
9388        other   NaN  NaN     95674          NaN          NaN       NaN   

     description state  lat  long posting_date price_category date  
6728         NaN    nh  NaN   NaN          NaN            low  NaN  
9388         NaN    wa  NaN   NaN          NaN         medium  NaN

---

можно их удалить и сохранить в новый датафрейм df_without_nan

---

In [177]:
df_without_nan = df.dropna(subset = ['date']).copy()

In [178]:
df_without_nan.shape

(9998, 20)

---

*year имеет тип float64 вместо int*

*posting_date, date имеют тип object вместо data*

*изменим тип данных*

*df['year'] = df['year'].astype('int') пока не можем сделать, так как есть пропуски*

---

In [179]:
df_without_nan['date'] = pd.to_datetime(df_without_nan.date, dayfirst=True)
#df['posting_date'] = pd.to_datetime(df.posting_date, dayfirst=True)

In [180]:
df_without_nan.date.dtypes

datetime64[ns, UTC]

---

*posting_date не изменил тип, так как выглядит*

*2021-04-17T12:30:50-0400 вместо*

*2021-04-17 16:30:50+00:00* *           то есть имеет неправильную структуру*

*2021-04-17 12:30:50+04:00*

*2021-04-17 12:30:50+04:00*



---

In [181]:
# приведем posting_date в читабельный для datetime64 вид
i = 0
for value in df_without_nan['posting_date']:
    x = str(value)
    value = x.replace('T', ' ') # вместо "Т" заменяем " "
    result = value[:19] + '+' + value[20:22] + ':' + value[22:] # переводим 12:30:50-0400 в такой вид 12:30:50+04:00
    df.at [i, 'posting_date'] = result # запишем отформатированное значение в датафрейм
    i = i + 1
   
    
    

In [182]:
df_without_nan[['date', 'posting_date']]

date               posting_date
0    2021-04-17 16:30:50+00:00  2021-04-17 12:30:50+04:00
1    2021-05-03 19:40:21+00:00  2021-05-03 15:40:21+04:00
2    2021-04-28 10:52:20+00:00  2021-04-28 03:52:20+07:00
3    2021-04-17 14:08:57+00:00  2021-04-17 10:08:57+04:00
4    2021-05-03 22:32:06+00:00  2021-05-03 18:32:06+04:00
...                        ...                        ...
9995 2021-04-10 20:33:57+00:00  2021-04-22 12:14:01+07:00
9996 2021-05-03 13:36:30+00:00  2021-04-14 09:14:42+05:00
9997 2021-04-22 19:14:01+00:00  2021-04-24 13:50:49+04:00
9998 2021-04-14 14:14:42+00:00   2021-04-14T09:14:42-0500
9999 2021-04-24 17:50:49+00:00   2021-04-24T13:50:49-0400

[9998 rows x 2 columns]

In [183]:
df_without_nan['posting_date'] = pd.to_datetime(df_without_nan.posting_date, dayfirst=True)

In [184]:
df_without_nan['posting_date'].dtypes

dtype('O')

### Handling missing values

### Anomaly processing

3. *Feature engineering*

Объявите блок Feature engineering. В этом блоке:
* подготовьте категориальные переменные с помощью OneHotEncoder;
* стандартизируйте и нормализуйте переменные, если нужно;
* создайте новые признаки на основе информации в датафрейме (на основе дат, текстовых значений переменных, и так далее);
* удалите неинформативные колонки, которые появились в датасете в результате Feature engineering;
* сформируйте финальный датасет, на котором будет производиться моделирование, и сохраните его в отдельный файл.

В решении данной задачи опирайтесь на результаты работы в 20 модуле блока Data Preparation. Исследование подкрепляйте соответствующей визуализацией в виде графиков, таблиц или матриц корреляций, если это нужно.

Декомпозируйте блок на подблоки.

4. *Modelling*

Объявите блок Modelling. В этом блоке:
* сформируйте датасет для обучения; инициализируйте фичи и целевую переменную 
(`price_category`); положите их в соответствующие переменные;
* разделите данные на треин и тест;
* объявите три модели: логистическая регрессия, случайный лес и многослойный персептрон;
* поработайте с моделями: сделайте тюнинг параметров и выберите лучшую модель с помощью кросс-валидации на тренировочной выборке;
* по результатам кросс-валидации выберите лучшую модель;
* посчитайте значение метрики лучшей модели на тестовой выборке; убедитесь, что модель не переобучилась.

В решении данной задачи опирайтесь на результаты работы в 22-23 модулях блока Modelling. Исследование подкрепляйте соответствующей визуализацией в виде графиков, таблиц или матриц корреляций, если это нужно.

Декомпозируйте блок на подблоки, если это нужно.

4. *Results*

Объявите блок Results. В этом блоке:

* подведите итог: какая модель показала себя лучше всего и будет финальным результатом данного ноутбука;
* обучите эту модель на всём датасете;
* сохраните обученную модель в `pickle`.

## Как отправить работу на проверку


Отправьте куратору на проверку финальный ноутбук проекта, а также файл обученной модели в формате `pickle`.